src/main/java/net/minecraftforge/common/data/SoundDefinitionsProvider.java

```diff
@@ -0,0 +1,270 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.data;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonObject;
+import net.minecraft.data.DataGenerator;
+import net.minecraft.data.DirectoryCache;
+import net.minecraft.data.IDataProvider;
+import net.minecraft.resources.ResourcePackType;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.SoundEvent;
+import net.minecraftforge.registries.ForgeRegistries;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * Data provider for the {@code sounds.json} file, which identifies sound definitions
+ * for the various sound events in Minecraft.
+ */
+public abstract class SoundDefinitionsProvider implements IDataProvider
+{
+    private static final Logger LOGGER = LogManager.getLogger();
+    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
+    private final DataGenerator generator;
+    private final String modId;
+    private final ExistingFileHelper helper;
+
+    private final Map<String, SoundDefinition> sounds = new LinkedHashMap<>();
+
+    /**
+     * Creates a new instance of this data provider.
+     *
+     * @param generator The data generator instance provided by the event you are initializing this provider in.
+     * @param modId The mod ID of the current mod.
+     * @param helper The existing file helper provided by the event you are initializing this provider in.
+     */
+    protected SoundDefinitionsProvider(final DataGenerator generator, final String modId, final ExistingFileHelper helper)
+    {
+        this.generator = generator;
+        this.modId = modId;
+        this.helper = helper;
+    }
+
+    /**
+     * Registers the sound definitions that should be generated via one of the {@code add} methods.
+     */
+    public abstract void registerSounds();
+
+    @Override
+    public void act(DirectoryCache cache) throws IOException
+    {
+        this.sounds.clear();
+        this.registerSounds();
+        this.validate();
+        if (!this.sounds.isEmpty())
+        {
+            this.save(cache, this.generator.getOutputFolder().resolve("assets/" + this.modId + "/sounds.json"));
+        }
+    }
+
+    @Override
+    public String getName()
+    {
+        return "Sound Definitions";
+    }
+
+    // Quick helpers
+    /**
+     * @see SoundDefinition#definition()
+     */
+    protected static SoundDefinition definition()
```

Comment:
It seems a bit unnecessary to define bouncers for all these methods when they could just be defined in this class entirely. What purpose does defining them inside the implementation serve?

Reply:
I defined the factory methods in the implementation classes so that everything related to them would be in the same place, and to keep the constructors private. These bouncer methods exist mainly as a convenience, so that people wouldn't need to specify the class every time they want to create a new sound or a new definition, without using `static` imports.

Reply:
Why do the constructors need to be private? They aren't for most other data builder classes.

Reply:
You are correct in general, and maybe I'm in the minority here, but I feel like when you have to start duplicating methods in two different places just to make your design work, that design has failed. Sometimes you have to compromise on what is theoretically right vs what is easier to write and maintain. Mostly these data builder classes will never be directly instantiated, even if the constructors are public. If we need to change the ctor to a static factory later, it can be done with little API interruption between versions. If you want to leave a *single* static factory on the class and just move all the overloads out, that would also be fine.

Reply:
Fair enough; will do.